File: apps/web/lib/services/recommendation.service.ts
import {
  getCachedRecommendations,
  setCachedRecommendations,
  getUserProfile,
  getProductViewsByUser,
  getMostViewedProducts,
  getMostCartedProducts,
  getMostWishlistedProducts,
} from "@workspace/db/queries/index"
import { getProducts } from "@workspace/db/queries/product"
import type {
  RecommendedProduct,
  RecommendationStrategy,
} from "../types/recommendations"

const AI_SERVICE_URL =
  process.env.AI_SERVICE_URL ?? "http://localhost:8000"

export class RecommendationService {
  async getRecommendations(
    userId: string,
    limit = 20,
    context?: Record<string, unknown>
  ): Promise<RecommendedProduct[]> {
    // 1. Check cache first
    const cached = await getCachedRecommendations(userId, "hybrid")
    if (cached) {
      return cached.recommendations
        .slice(0, limit)
        .map((r) => ({
          ...r,
          strategy: "hybrid" as RecommendationStrategy,
        }))
    }

    // 2. Get user profile to determine strategy
    const profile = await getUserProfile(userId)

    let recommendations: RecommendedProduct[]

    if (!profile || (profile.totalViews ?? 0) < 5) {
      // Cold start: use popular + rule-based
      recommendations = await this.getColdStartRecommendations(
        limit,
        context
      )
    } else {
      // Warm user: try ML service, fallback to local hybrid
      recommendations = await this.getHybridRecommendations(
        userId,
        limit,
        context
      )
    }

    // 3. Cache results
    if (recommendations.length > 0) {
      await setCachedRecommendations({
        userId,
        strategy: "hybrid",
        recommendations: recommendations.map((r) => ({
          productId: r.productId,
          score: r.score,
          reason: r.reason,
        })),
        context,
        ttlMinutes: 30,
      }).catch(() => {
        // Cache failure is non-critical
      })
    }

    return recommendations
  }

  async getSimilarProducts(
    productId: string,
    limit = 10
  ): Promise<RecommendedProduct[]> {
    // Try ML service first
    try {
      const response = await fetch(
        `${AI_SERVICE_URL}/similar-products/${productId}?limit=${limit}`,
        { signal: AbortSignal.timeout(5000) }
      )
      if (response.ok) {
        const data = (await response.json()) as {
          recommendations: RecommendedProduct[]
        }
        return data.recommendations
      }
    } catch {
      // ML service unavailable, fallback to category-based
    }

    return this.getCategorySimilar(productId, limit)
  }

  private async getColdStartRecommendations(
    limit: number,
    context?: Record<string, unknown>
  ): Promise<RecommendedProduct[]> {
    const [popular, trending] = await Promise.all([
      getMostViewedProducts(limit),
      getMostCartedProducts(limit),
    ])

    const productIds = [
      ...new Set([
        ...popular.map((p) => p.productId),
        ...trending.map((p) => p.productId),
      ]),
    ].slice(0, limit)

    const allProducts = await getProducts({
      category: context?.category as string | undefined,
    })
    const productMap = new Map(allProducts.map((p) => [p.id, p]))

    return productIds
      .filter((id) => productMap.has(id))
      .map((id, i) => ({
        productId: id,
        score: 1 - i * 0.05,
        reason: "Trending and popular",
        strategy: "popular" as RecommendationStrategy,
      }))
  }

  private async getHybridRecommendations(
    userId: string,
    limit: number,
    context?: Record<string, unknown>
  ): Promise<RecommendedProduct[]> {
    // Try ML service
    try {
      const response = await fetch(
        `${AI_SERVICE_URL}/recommend/${userId}`,
        {
          method: "POST",
          headers: { "Content-Type": "application/json" },
          body: JSON.stringify({ limit, context }),
          signal: AbortSignal.timeout(5000),
        }
      )
      if (response.ok) {
        const data = (await response.json()) as {
          recommendations: RecommendedProduct[]
        }
        return data.recommendations
      }
    } catch {
      // ML service unavailable
    }

    // Fallback: local rule-based scoring
    return this.getLocalHybridRecommendations(userId, limit, context)
  }

  private async getLocalHybridRecommendations(
    userId: string,
    limit: number,
    _context?: Record<string, unknown>
  ): Promise<RecommendedProduct[]> {
    const [profile, views, wishlisted] = await Promise.all([
      getUserProfile(userId),
      getProductViewsByUser(userId, 100),
      getMostWishlistedProducts(limit),
    ])

    const allProducts = await getProducts({})
    const viewedSet = new Set(views.map((v) => v.productId))
    const wishlistedSet = new Set(
      wishlisted.map((w) => w.productId)
    )

    // Score each product
    const scored = allProducts
      .filter((p) => !viewedSet.has(p.id)) // exclude already viewed
      .map((p) => {
        let score = 0

        // Category affinity scoring
        if (profile?.categoryAffinities) {
          const affinity =
            (
              profile.categoryAffinities as Record<string, number>
            )[p.category] ?? 0
          score += affinity * 3
        }

        // Price fit scoring
        if (profile?.preferredPriceRange) {
          const range = profile.preferredPriceRange as {
            min: number
            max: number
          }
          const price = Number(p.price)
          if (price >= range.min && price <= range.max) {
            score += 2
          }
        }

        // Wishlist boost
        if (wishlistedSet.has(p.id)) {
          score += 1.5
        }

        // Rating boost
        if (p.rating) {
          score += Number(p.rating) * 0.5
        }

        // Recency boost
        const daysSinceCreated =
          (Date.now() - new Date(p.createdAt).getTime()) / 86400000
        if (daysSinceCreated < 7) score += 1
        else if (daysSinceCreated < 30) score += 0.5

        return {
          productId: p.id,
          score: Math.round(score * 100) / 100,
          reason: this.generateReason(p, profile),
          strategy: "rule_based" as RecommendationStrategy,
        }
      })
      .sort((a, b) => b.score - a.score)
      .slice(0, limit)

    return scored
  }

  private async getCategorySimilar(
    productId: string,
    limit: number
  ): Promise<RecommendedProduct[]> {
    const allProducts = await getProducts({})
    const target = allProducts.find((p) => p.id === productId)
    if (!target) return []

    return allProducts
      .filter(
        (p) =>
          p.id !== productId && p.category === target.category
      )
      .slice(0, limit)
      .map((p, i) => ({
        productId: p.id,
        score: 1 - i * 0.1,
        reason: `Similar to ${target.name}`,
        strategy: "content_based" as RecommendationStrategy,
      }))
  }

  private generateReason(
    product: { category: string; rating: string | null },
    profile: { categoryAffinities: unknown } | null | undefined
  ): string {
    const affinities = profile?.categoryAffinities as
      | Record<string, number>
      | undefined
    if (affinities && (affinities[product.category] ?? 0) > 0.5) {
      return `Based on your interest in ${product.category}`
    }
    if (product.rating && Number(product.rating) >= 4) {
      return "Highly rated product"
    }
    return "Recommended for you"
  }
}

export const recommendationService = new RecommendationService()

File: apps/ai/app/services/recommendation_service.py
import json
import logging
from typing import Any, Dict, Optional

import numpy as np

from app.models.schemas import RecommendedProduct
from app.services.embedding_service import embedding_service
from app.utils.database import (
    fetch_product_by_id,
    fetch_similar_products_by_vector,
    fetch_top_cooccurrence_products,
    fetch_products,
    fetch_user_events,
    fetch_user_profile,
)

logger = logging.getLogger(__name__)


class RecommendationService:
    """Hybrid recommendation engine combining multiple strategies."""

    def recommend(
        self,
        user_id: str,
        limit: int = 20,
        context: Optional[Dict] = None,
    ) -> tuple[list[RecommendedProduct], str]:
        """Generate recommendations for a user.

        Returns (recommendations, strategy_used).
        """
        profile = fetch_user_profile(user_id)
        events = fetch_user_events(user_id, limit=500)

        total_views = profile.get("total_views", 0) if profile else 0

        if total_views < 5:
            # Cold start
            recs = self._cold_start(limit, context)
            return recs, "popular"

        # Warm user: hybrid approach
        recs = self._hybrid_recommend(
            user_id, profile, events, limit, context
        )
        return recs, "hybrid"

    def similar_products(
        self, product_id: str, limit: int = 10
    ) -> list[RecommendedProduct]:
        """Find similar products using embeddings + category matching."""
        product = fetch_product_by_id(product_id)
        if not product:
            return []

        # Try embedding-based similarity
        product_text = embedding_service.build_product_text(product)
        query_embedding = embedding_service.encode(product_text)
        vector_matches = fetch_similar_products_by_vector(
            query_embedding, limit=limit + 1
        )
        vector_recs = self._vector_matches_to_recommendations(
            product_id, vector_matches, limit
        )
        if vector_recs:
            return vector_recs

        # Fallback: category-based
        return self._category_similarity(product, limit)

    def search_rerank(
        self,
        query: str,
        product_ids: list[str],
        user_id: Optional[str] = None,
        limit: int = 20,
    ) -> tuple[list[str], list[float]]:
        """Rerank search results using semantic similarity."""
        query_embedding = embedding_service.encode(query)

        products = fetch_products(limit=1000)
        product_map = {str(p["id"]): p for p in products}

        scored: list[tuple[str, float]] = []
        for pid in product_ids:
            p = product_map.get(pid)
            if not p:
                continue

            text = embedding_service.build_product_text(p)
            product_embedding = embedding_service.encode(text)
            sim = embedding_service.cosine_similarity(
                query_embedding, product_embedding
            )

            # Boost by rating
            rating = float(p.get("rating") or 0)
            score = sim * 0.7 + (rating / 5.0) * 0.3

            scored.append((pid, score))

        scored.sort(key=lambda x: x[1], reverse=True)
        scored = scored[:limit]

        return [s[0] for s in scored], [s[1] for s in scored]

    def _cold_start(
        self, limit: int, context: Optional[Dict] = None
    ) -> list[RecommendedProduct]:
        """Popular + rule-based recommendations for new users."""
        products = fetch_products(limit=200)

        scored: list[tuple[dict, float]] = []
        for p in products:
            score = 0.0

            # Rating boost
            rating = float(p.get("rating") or 0)
            score += rating * 2

            # Review count boost (popularity)
            reviews = int(p.get("review_count") or 0)
            score += min(reviews / 100, 2.0)

            # Context matching
            if context:
                score += self._context_score(p, context)

            scored.append((p, score))

        scored.sort(key=lambda x: x[1], reverse=True)

        return [
            RecommendedProduct(
                product_id=str(p["id"]),
                score=round(s, 3),
                reason="Popular and highly rated",
                strategy="popular",
            )
            for p, s in scored[:limit]
        ]

    def _hybrid_recommend(
        self,
        user_id: str,
        profile: Optional[dict],
        events: list[dict],
        limit: int,
        context: Optional[Dict] = None,
    ) -> list[RecommendedProduct]:
        """Combine content-based, collaborative, and rule-based signals."""
        products = fetch_products(limit=500)
        collaborative_candidates = fetch_top_cooccurrence_products(
            user_id, limit=200
        )
        collaborative_map = {
            str(item["product_id"]): float(item["score"])
            for item in collaborative_candidates
            if item.get("product_id")
        }

        # Products user has already interacted with
        interacted_ids = set()
        for e in events:
            if e.get("product_id"):
                interacted_ids.add(str(e["product_id"]))

        candidates = [
            p for p in products if str(p["id"]) not in interacted_ids
        ]

        scored: list[tuple[dict, float, str]] = []
        for p in candidates:
            score = 0.0
            reasons: list[str] = []

            # 1. Category affinity (content-based)
            if profile and profile.get("category_affinities"):
                affinities = profile["category_affinities"]
                if isinstance(affinities, str):
                    affinities = json.loads(affinities)
                cat_score = float(affinities.get(p["category"], 0))
                score += cat_score * 3
                if cat_score > 0.3:
                    reasons.append(
                        f"Matches your interest in {p['category']}"
                    )

            # 2. Price fit
            if profile and profile.get("preferred_price_range"):
                pr = profile["preferred_price_range"]
                if isinstance(pr, str):
                    pr = json.loads(pr)
                price = float(p.get("price") or 0)
                if pr.get("min", 0) <= price <= pr.get("max", 99999):
                    score += 2
                    reasons.append("Within your price range")

            # 3. Occasion match
            if profile and profile.get("top_occasions"):
                occ = profile["top_occasions"]
                if isinstance(occ, str):
                    occ = json.loads(occ)
                product_occasions = p.get("occasions") or []
                if isinstance(product_occasions, str):
                    product_occasions = json.loads(product_occasions)
                for o in occ[:3]:
                    occ_name = o.get("occasion", "") if isinstance(o, dict) else str(o)
                    if occ_name in product_occasions:
                        score += 4
                        reasons.append(f"Great for {occ_name}")
                        break

            # 4. Rating boost
            rating = float(p.get("rating") or 0)
            score += rating * 0.5

            # 5. Recency boost
            score += 1  # base score for availability

            # 6. Collaborative boost
            cooccurrence = collaborative_map.get(str(p["id"]), 0.0)
            if cooccurrence > 0:
                score += min(cooccurrence, 5.0)
                reasons.append("Users with similar behavior liked this")

            # 7. Context matching
            if context:
                ctx_score = self._context_score(p, context)
                score += ctx_score

            reason = reasons[0] if reasons else "Recommended for you"
            scored.append((p, score, reason))

        scored.sort(key=lambda x: x[1], reverse=True)

        return [
            RecommendedProduct(
                product_id=str(p["id"]),
                score=round(s, 3),
                reason=reason,
                strategy="hybrid",
            )
            for p, s, reason in scored[:limit]
        ]

    def _context_score(self, product: dict, context: dict) -> float:
        """Score product against request context."""
        score = 0.0

        if context.get("occasion"):
            occasions = product.get("occasions") or []
            if isinstance(occasions, str):
                occasions = json.loads(occasions)
            if context["occasion"] in occasions:
                score += 4

        if context.get("recipientType"):
            recipients = product.get("recipient_types") or []
            if isinstance(recipients, str):
                recipients = json.loads(recipients)
            if context["recipientType"] in recipients:
                score += 3

        if context.get("budget"):
            price = float(product.get("price") or 0)
            budget = context["budget"]
            if isinstance(budget, dict):
                if budget.get("min", 0) <= price <= budget.get(
                    "max", 99999
                ):
                    score += 2

        if context.get("category"):
            if product.get("category") == context["category"]:
                score += 3

        return score

    def _vector_matches_to_recommendations(
        self, product_id: str, vector_matches: list[dict], limit: int
    ) -> list[RecommendedProduct]:
        filtered = [
            row
            for row in vector_matches
            if str(row["product_id"]) != str(product_id)
        ][:limit]
        return [
            RecommendedProduct(
                product_id=str(row["product_id"]),
                score=round(float(row["score"]), 3),
                reason="Similar product",
                strategy="content_based",
            )
            for row in filtered
        ]

    def _category_similarity(
        self, product: dict, limit: int
    ) -> list[RecommendedProduct]:
        """Fallback: find products in the same category."""
        products = fetch_products(limit=200)

        same_category = [
            p
            for p in products
            if p["category"] == product["category"]
            and str(p["id"]) != str(product["id"])
        ]

        return [
            RecommendedProduct(
                product_id=str(p["id"]),
                score=round(1.0 - i * 0.05, 3),
                reason=f"Similar to {product.get('name', 'this product')}",
                strategy="content_based",
            )
            for i, p in enumerate(same_category[:limit])
        ]


recommendation_service = RecommendationService()
